### PR TITLE
Validator

### PR DESCRIPTION
**Pull request by Drgy**

### CODE REVIEW

**Review by andreogle — CHANGES_REQUESTED** [review]
I really like the idea of making this a flexible package that has a declarative way to validate a JSON file against a given schema. This will hopefully make it really easy to adjust when changes are made to the OIS. I was expecting a mostly hardcoded package so this is much better than that. Eventually, this could probably even become a stand alone package that Airnode just makes use of to validate JSON files 👍🏻 

I'm not really a fan of the current way the template schemas work though. Parts of it feels a bit awkward to me (particularly things like the `if`/`then`). I would rather it be a bit more explicit but I'm not really sure how to improve it at the moment. Things like validating sibling and children keys makes it complicated.

It would also be nice to have this runnable as a script. A nice UX could be something like (or however you pass in arguments to NPM scripts):

```sh
npm run validate:api-specs --file config.json --template templates/api.json
``` 
Not important now but just something to keep in mind that might influence how it's built.

I've left some comments around refactoring. I spent quite a lot of time trying to understand everything but still didn't manage to get it all. I'm hoping that if we can refactor to improve the readability and maintainability then I'll try review again. If something doesn't make sense, please let me know.

**Comment by Drgy** [comment]
Thanks for the feedback, I spent a lot of time with testing and trying out various versions of the keywords/structures and the ones that are used the most (like `if`, `then`, `arrayObject`...) are the best I could think of, if you would have anything specific, I can definitely try that, but I don't think I will come up with something better. The one thing I'm not quite happy with is the absolute/relative validation, in the conditions its not quite clear whats being evaluated I think, this part was kind of hacked in, after I found out it will be needed.

Adding npm script is a good idea, I think thats going to be useful, I will certainly add that. Also I noticed there is a new directory `packages/ois/src/validator` which seems is intended for validator, should I move validator there or keep it where it is?

I see how the code might be confusing, I was working on it for quite a while and some fixes for small bugs are hacked in a way thats not ideal, which might add unwanted obfuscation to the code. I will try to simplify some sections and leave comments where the intentions might not be clear at first.

File: packages/validator/__tests__/validator.test.js [hunk 1]
```diff
@@ -0,0 +1,797 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/__tests__/validator.test.js]: Moved in f568eaf

File: packages/validator/__tests__/validator.test.js [hunk 2]
```diff
@@ -0,0 +1,797 @@
...
+    ]
+  }
+]`;
+
+/* OIS SPECS */
+const specs19 = `{
+"oisFormat": "1.10.0",
+"title": "myOisTitle",
+"version": "1.2.3",
+"apiSpecifications": ${specs1},
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/__tests__/validator.test.js]: Added in 62eb558

File: packages/validator/__tests__/validator.test.js [hunk 3]
```diff
@@ -0,0 +1,797 @@
...
+
+    test('multiple array items', () => {
+      expect(validator.isApiSpecsValid(specs5)).toMatchObject({ valid: true, messages: [] });
+      expect(validator.isApiSpecsValid(specs6)).toMatchObject({
+        valid: false,
+        messages: [sizeExceededMessage('servers', 1)],
+      });
+    });
+
+    test('edge cases', () => {
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/__tests__/validator.test.js]: Renamed in 5e852ff, I hope this would be enough, each group of tests is focused on one specific aspect of validator, which can be described using only few words, something more descriptive would probably mean writing out what exactly it is testing, which would take multiple lines and imho make it not easy to understand what the section is testing at first glance.

File: packages/validator/__tests__/validator.test.js [hunk 4]
```diff
@@ -0,0 +1,797 @@
+'use strict';
+
+const validator = require('../lib/validator');
+
+const specs1 = `{
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/__tests__/validator.test.js]: Renamed and moved in 62eb558. This is something I should have addressed sooner, at first I thought there would be just a few tests so it would be fine, but it quickly got out of hand... 
Each specification is now right above the expect it is used in, so there is minimal scrolling required to view them. 
I don't think one spec per test here is a good idea. For example in `missing parameters` test, there are multiple specs which are testing the same thing, however since requirements for parameters might be influenced by many things, there are multiple similar inputs, the only way to differentiate them would be write out the errors they should produce, but in that case I would much rather stick with the plain numbers in the same test. Other than that the specifications have a meaningful name, not the previous specs1, 2, 3...

**Comment by bbenligiray** [comment]
This being a separate package makes sense to me too. Would we move it out of the monorepo then @andreogle ?

**Comment by andreogle** [comment]
@Drgy I don't really have any concrete suggestions for how to improve the template layout at the moment. It's pretty clear that the requirements make it very complicated. There's definitely a bit of a learning curve with the current implementation, but I think that's fine for now.

Yeah the intention was for this to live in the `packages/ois` directory, but we can move it at any point really.

@bbenligiray Yeah I think the generic/core parts of it should probably be extracted into a separate repo, although that doesn't necessarily need to happen now. They aren't really airnode specific at all - only the "template specs" are. It might also encourage potential contributors to help out.

File: packages/validator/lib/validatorHelper.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+function getLastParamName(paramPath) {
+  const lastDotIndex = paramPath.lastIndexOf('.');
+  let paramName = paramPath;
+
+  if (lastDotIndex >= 0) {
+    paramName = paramPath.slice(lastDotIndex + 1);
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validatorHelper.js]: Fixed in 55c2d2d

File: packages/validator/lib/validatorHelper.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    let newKey = key.replace(/__match/g, match);
+    parsedSpecs[newKey] = typeof specs[key] === 'string' ? specs[key].replace(/__match/g, match) : replaceConditionalMatch(match, specs[key]);
+  }
+
+  return parsedSpecs;
+}
+
+function checkRedundancy(nonRedundant, specs, paramPath, messages) {
+  if (typeof specs === 'object') {
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validatorHelper.js]: Updated in 85be84a
In case the specs is array, regular `for` loop seems like the best solution. If `map` would be used the result needs flattened, which can be avoided by using `for`...

File: packages/validator/lib/validatorHelper.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+          }
+
+          messages.push({ level: 'warning', message: `Extra field: ${paramPath}${paramPath ? '.' : ''}${param}` });
+        }
+      }
+    }
+  }
+}
+
+function insertNonRedundantParam(param, specsStruct, nonRedundantParams, specs) {
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validatorHelper.js]: Simplified in 94a2343
non redundant is not the ideal name, but renaming it to required would suggest that there are only required parameters stored, but in fact there are required and optional variables (also variables that become required in case some conditions are met).

File: packages/validator/lib/validatorHelper.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+  if (lastDotIndex >= 0) {
+    paramName = paramPath.slice(lastDotIndex + 1);
+  }
+
+  return paramName;
+}
+
+function replaceConditionalMatch(match, specs) {
+  let parsedSpecs = {};
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validatorHelper.js]: Updated in a956d46

File: packages/validator/__tests__/validator.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreogle) [reply packages/validator/__tests__/validator.test.js]: Alright that's fine. I think the renaming already looks much better 👍🏻

File: packages/validator/__tests__/validator.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreogle) [reply packages/validator/__tests__/validator.test.js]: Renames look good to me

File: packages/validator/lib/validatorHelper.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreogle) [reply packages/validator/lib/validatorHelper.js]: Alright. I don't have any other suggestions at the moment either

File: packages/validator/lib/validator.js [hunk 9]
```diff
@@ -0,0 +1,424 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Renamed in df5028f

File: packages/validator/lib/validator.js [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+const fs = require('fs');
+const helper = require('./validatorHelper');
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Renamed in 494fece

File: packages/validator/lib/validator.js [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+                    let result = validateSpecs(specs[thisName], parsedSpecs, `${paramPath}${paramPath ? '.' : ''}${thisName}`, specsRoot, nonRedundantParams[thisName], nonRedundantParamsRoot, paramPathPrefix);
+
+                    if (!result.valid) {
+                      if (Object.keys(nonRedundantParamsCopy).length) {
+                        nonRedundantParams[thisName] = nonRedundantParamsCopy;
+                      } else {
+                        delete nonRedundantParams[thisName];
+                      }
+
+                      messages.push({ level: 'error', message: `Condition in ${paramPath}${paramPath ? '.' : ''}${thisName} is not met with ${param}` });
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Added in 773429e

File: packages/validator/lib/validator.js [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+                      valid = false;
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Updated in aa4d771

File: packages/validator/package.json [hunk 13]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "validator",
+  "version": "0.0.0",
+  "description": "Verifies that provided specs.json or security.json files are valid Airnode compatible configurations",
+  "author": "Drgy <drgonjak@gmail.com>",
+  "homepage": "https://github.com/clc-group/airnode#readme",
+  "license": "ISC",
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/package.json]: Yeah all of these were generated with lerna, back in the old repo
updated in 36225e0

File: packages/validator/lib/validator.js [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+      case '__conditions':
+        for (const condition of specsStruct[key]) {
+          if (condition['__if']) {
+            const paramName = Object.keys(condition['__if'])[0];
+            const paramValue = condition['__if'][paramName];
+            const thenParamName = Object.keys(condition['__then'])[0];
+
+            if (paramName === '__this') {
+              for (const thisName of Object.keys(specs)) {
+                if (!thisName) {
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Removed in 25ad4d0

File: packages/validator/lib/validator.js [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+
+function validateSpecs(specs, specsStruct, paramPath, specsRoot, nonRedundantParams, nonRedundantParamsRoot, paramPathPrefix = '') {
+  let messages = [];
+  let valid = true;
+  let tmpNonRedundant = [];
+  let tmpResult = {};
+
+  for (const key of Object.keys(specsStruct)) {
+    switch (key) {
+      case '__conditions':
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Updated in f00f517

File: packages/validator/lib/validator.js [hunk 16]
```diff
@@ -0,0 +1,424 @@
...
+              }
+            }
+          }
+        }
+
+        break;
+
+      case '__regexp':
+        if (typeof specs !== 'string' || !specs.match(new RegExp(specsStruct[key]))) {
+          let level = 'warning';
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Simplified in 35baf5b

File: packages/validator/lib/validator.js [hunk 17]
```diff
@@ -0,0 +1,424 @@
...
+
+        break;
+
+      case '__arrayItem':
+        if (!nonRedundantParams) {
+          nonRedundantParams = [];
+        }
+
+        for (let i = 0; i < specs.length; i++) {
+          nonRedundantParams.push({});
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: This part was validating arrays, if its the place I think it is. `nonRedundantParams` structure must be the same as structure of specs, so at the end we can check for extra parameters. In this case its necessary to push empty object into the array, so that when validateSpecs is recursively called few lines later, there is something that we can pass as `nonRedundantParams`. I hope that makes sense, I added some comments into the code, perhaps its more clear now.

File: packages/validator/lib/validator.js [hunk 18]
```diff
@@ -0,0 +1,424 @@
...
+  }
+
+  if (specs === specsRoot) {
+    helper.checkRedundancy(nonRedundantParamsRoot, specs, paramPath, messages);
+  }
+
+  return { valid, messages };
+}
+
+function isApiSpecsValid(specs) {
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/lib/validator.js]: Simplified in 8967b3a
`nonRedundant` variable must stay, because its passed by reference.

File: packages/validator/src/validator.test.js [hunk 19]
```diff
@@ -0,0 +1,884 @@
...
+            "securitySchemes": {}
+        },
+        "security": {}`;
+      expect(validator.isApiSpecsValid(invalidJSONAPISpec)).toMatchObject({
+        valid: false,
+        messages: [{ level: 'error', message: 'SyntaxError: Unexpected end of JSON input' }],
+      });
+    });
+
+    test('missing parameters', () => {
```
Comment (andreogle) [inline packages/validator/src/validator.test.js:463]: Just for consistency, I think let's use `it()` instead of `test()` for test cases - they do the exact same thing. All of the other packages in the monorepo use `it()`

**Review by andreogle — CHANGES_REQUESTED** [review]
This reads *significantly* easier after the refactor! Thank you for that 💯 💯

I've left a few more (mostly small) comments. One of the main points is about the functions mutating argument references directly, which I don't think is a great pattern to follow. I think it would be a bit better if we could rather return a new object that the calling function then uses as needed.

Also, I'm not really sure that ESLint is being run against this code. It should pick up when `let` is unused for variables that are not reassigned. Could you check that out and fix (or disable) any errors come up please?

I think we're nearly there with getting this merged

File: packages/validator/src/validator.js [hunk 20]
```diff
@@ -0,0 +1,576 @@
...
+        nonRedundantParams[nonRedundantParams.length - 1],
+        roots
+      );
+
+      if (!result.messages.length) {
+        return true;
+      }
+
+      nonRedundantParams[paramIndex] = nonRedundantParamsCopy;
+    }
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/src/validator.js:41]: Added in 527ce4a,
couldnt think of a scenario when reading the condition, matching with `if` is not necessary though...

[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/src/validator.js:41]: Updated in 978877a

File: packages/validator/src/validator.js [hunk 21]
```diff
@@ -0,0 +1,576 @@
...
+
+    if (!keepRedundantParams) {
+      // returning nonRedundantParams to original state, since it wasn't used in "then section" of condition
+      if (Object.keys(nonRedundantParamsCopy).length) {
+        nonRedundantParams[thenParamName] = nonRedundantParamsCopy;
+      } else {
+        delete nonRedundantParams[thenParamName];
+      }
+    }
+  } else if (thenParamName === '__any') {
```
[the remark this answers is not in the document]
Comment (Drgy) [reply packages/validator/src/validator.js]: Updated in 978877a

File: packages/validator/src/validator.test.js (hunk 19 above)
Comment (Drgy) [reply packages/validator/src/validator.test.js]: Replaced in e678ec5

**Review by andreogle — APPROVED** [review]
Looks great now that it's in TS 👌🏻 . I also like the JSDoc descriptions (even for the internal functions) - they help to understand the code. I don't have any further comments on the actual code parts for now so I'm happy with that.

I do think that we should re-arrange the tests a bit. I think most of the code branches are probably covered by tests, but now since you've managed to break these functions into separate files, it could be much simpler to test them.

Right now, the `validator` tests are tightly coupled to each of these modules. So, if something was to change in one of the validator modules, many other (bigger) tests elsewhere would also break. Ideally only the tests in that specific module should break.

I think extracting the tests to the individual modules should make them significantly smaller and easier to maintain. Testing everything in an 882 line validator test file isn't really ideal. Test coverage is definitely not the best metric to evaluate a code base by, but it does give some useful indications. Having smaller tests makes it much easier to get better coverage.

Let me know what you think.
